### PR TITLE
fix and deprecate dict() function

### LINKED ISSUE #311: TypeError when calling dict() method on inventory group
Python version: `3.7.1`
Nornir version: `2.0.0`
Config:
```
---
core:
  num_workers: 10
  raise_on_error: true
inventory:
  plugin: nornir.plugins.inventory.simple.SimpleInventory
  options:
    host_file: "hosts.yaml"
    group_file: "groups.yaml"
```
Hosts:
```
---
localhost:
  hostname: localhost
  platform: linux
  groups:
      - global
  data:
    site: test
    role: container
```
Groups:
```
---
global:
  port: 2222
  data:
    domain: example.com
```
Result:
```
In [1]: from nornir import InitNornir

In [2]: nr = InitNornir(config_file="config.yaml", dry_run=True)

In [3]: nr.inventory.hosts
Out[3]: {'localhost': Host: localhost}

In [4]: nr.inventory.groups
Out[4]: {'global': Group: global}

In [5]: nr.inventory.hosts['localhost'].dict()
Out[5]:
{'name': 'localhost',
 'connections': {},
 'defaults': <nornir.core.inventory.Defaults at 0x106f9c588>,
 'groups': ['global'],
 'data': {'site': 'test', 'role': 'container'},
 'connection_options': {},
 'hostname': 'localhost',
 'port': None,
 'username': None,
 'password': None,
 'platform': 'linux'}

In [6]: nr.inventory.groups['global'].dict()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-6cbbd72f0353> in <module>
----> 1 nr.inventory.groups['global'].dict()

~/repos/nornir/venv/lib/python3.7/site-packages/nornir/core/inventory.py in dict(self)
     34
     35     def dict(self):
---> 36         return {k: object.__getattribute__(self, k) for k in self.__recursive_slots__()}
     37
     38

~/repos/nornir/venv/lib/python3.7/site-packages/nornir/core/inventory.py in __recursive_slots__(self)
     28         for b in self.__class__.__bases__:
     29             if hasattr(b, "__recursive_slots__"):
---> 30                 s += b().__recursive_slots__()
     31             elif hasattr(b, "__slots__"):
     32                 s += b.__slots__

TypeError: __init__() missing 1 required positional argument: 'name'
```

### PR DESCRIPTION
**Pull request by dbarrosop**
Fixes #311

### CODE REVIEW

File: nornir/core/inventory.py [hunk 1]
```diff
@@ -23,17 +25,14 @@ def __init__(
...
-
     def dict(self):
-        return {k: object.__getattribute__(self, k) for k in self.__recursive_slots__()}
+        w = f"{self.__class__.__name__}.dict is deprecated, use nornir.core.deserializer instead"
+        warnings.warn(w)
+        return (
+            getattr(deserializer.inventory, self.__class__.__name__)
+            .serialize(self)
+            .dict()
+        )
```
Comment (dmfigol) [inline nornir/core/inventory.py]: Some nit-picking:
```
w = f"{self.dict.__qualname__} is deprecated ..."
```
https://www.python.org/dev/peps/pep-3155/

**Comment by dbarrosop** [comment]
@ktbyers @dmfigol is this one good? I'd like to merge and release soon

**Comment by dmfigol** [comment]
LGTM

**Comment by dbarrosop** [comment]
Thanks!

**Pull request merged**